src/rulesets/Base/base.jl

```diff
@@ -45,7 +45,7 @@
 @scalar_rule(-(x, y), (One(), -1))
 @scalar_rule(/(x, y), (inv(y), -(x / y / y)))
 @scalar_rule(\(x, y), (-(y / x / x), inv(x)))
-@scalar_rule(^(x, y), (y * x^(y - 1), Ω * log(x)))
+@scalar_rule(^(x, y), (ifelse(iszero(y), Zero(), y * x^(y - 1)), Ω * log(x)))
```

Comment:
This just seems like it unnecessarily introduces type instabilities. Is there a particular reason you need this?

Reply:
A better way to do this would be to overload `Base.literal_pow`. Have a look at how it is implemented in Julia base, it seems to be exactly for purposes like this.

Reply:
I thought this type instability is not a big deal because of union splitting? Also it seems to be the convention everywhere in ChainRules to return `Zero()` instead of trying to figure out the right zero in the rules.

I'll defer to @YingboMa about `Base.literal_pow`

Reply:
Small union optimization is better, but still has some slowdown.

Important thing about `Zero()`  is it's ability to remove `Thunks` via multiplication before it is evaluated.
Idk how often that occurs.

I think this this is a bit of a premature optimisation unless you have benchmarks showing it helps.

Reply:
Also note that there is no promotion mechanism for `Zero` and `One` yet, so user code might break if an unexpected `Zero` or `One` is returned

Reply:
That is not a problem. One should always be calling `extern` if you are not prepared for a Differential to come back.

Reply:
We need this branch, otherwise, exponentiation of zero is arguably wrong. See https://github.com/JuliaDiff/ForwardDiff.jl/commit/123cc19f081ac83c38b27d2c7a5e141217717fa3

Reply:
Ah, ok, I see. So the problem is that integers to a negative integer power throw a domain error. But for the purpose of type stability, `Zero()` should probably be replaced by `zero(x*y)`.

Reply:
`Zero()` is the convention in the rest of the package. But we can surely make an exception and forget about it.

Reply:
I am happy to go with `Zero`

Reply:
I've had another look into this and am worrying that this introduced type instability will cause significant overhead and bite us in the long run. I would imagine that an exponent of zero is really an edge case and won't occur too often in real code, so I'm wondering if not materializing that thunk in these cases is worth the trade-off of a significant performance hit in all other cases. Some quick benchmarks imply an up to a 25x(!) penalty with this instability (if I have overlooked something in my benchmarking methodology, please do correct me):
```julia
julia> f1(x, y) = ifelse(iszero(y), Zero(), y * x^(y - 1));

julia> f2(x, y) = ifelse(iszero(y), zero(x^y), y * x^(y - 1));

julia> a = rand(1000);

julia> @btime f1.($a, 2);
  24.201 μs (5 allocations: 8.05 KiB)

julia> @btime f2.($a, 2);
  787.900 ns (1 allocation: 7.94 KiB)
```

Reply:
Seems like we should change to `zero` and add a comment?

Reply:
@shashi are you happy to do that?

Reply:
What about
```
@scalar_rule(^(x, y), (ifelse(iszero(y), zero(Ω), y * x^(y - 1)), Ω * log(x)))
```
?

Floating-point exponentiation is very expensive.

Reply:
Yeah, I agree that that is legit.

(For my own futre reference:
```
julia> NaN^0.0
1.0
```
so this is correct, even then.
)

Reply:
I'm ok with `zero` or anything that is fast !